docs/dynamic-component-loading.md

```diff
@@ -0,0 +1,89 @@
+
+```csharp
+class AssemblyLoadContext
+{
+    public static AssemblyLoadContext CreateForAssemblyWithDependencies(string assemblyPath);
+
+    public static AssemblyLoadContext CreateForAssemblyWithDependencies(
+        string assemblyPath,
+        AssemblyLoadContext fallbackContext,
+        bool enableUnloading);
+}
+```
+
+"Advanced" version which would return an ALC instance and not just the main assembly. Allows for additional changes to the `AssemblyLoadContext`, like registering an event handler to the `Resolving` event for example.  
+The second overload adds the ability to specify:
+* `fallbackContext` which is the `AssemblyLoadContext` to defer to when the assembly resolution is not possible in this current context. By default this is the `AssemblyLoadContext.Default` (the app's context). This allows for creating effectively parent-child relationships between the load contexts.
+* `enableUnloading` which will mark the newly create load context to track references and to trigger unload fo the load context and all the assemblies in it when possible.
+
+## High-level description of the implementation
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication).
+* If the component has `.runtimeconfig.json` it will only be used to verify runtime version and provide probing paths.
+* The load context will determine a list of assemblies similar to TPA and list of native search paths. These will be used to resolve any assembly binding events in the load context.
+* If the load context can't resolve an assembly bind event, it will fallback to the parent load context (the app)
+
+## Important implications and limitations
+* Only framework dependent components will be supported. Self-contained components will not be supported even if there was a way to produce them.
+* The host (the app) can be any configuration (framework dependent or self-contained). The notion of frameworks is completely ignored by this new functionality.
+* All framework dependencies of the component must be resolvable by the app - simply put, the component must use the same frameworks as the app.
+* Components can't add frameworks to the app - the app must "pre-load" all necessary frameworks.
```

Comment:
Is this a strict requirement for all frameworks?  With the possibility of multiple frameworks this seems like an unnecessary restriction.  It might just work.  Is this just for restricting the set of supported and tested features?  

Perhaps it is reasonable to defer this until it becomes an issue.

Reply:
There are some small issues technically, but nothing major. This is more in line of what we want to support for now. The problem is that frameworks might not expect to exist in multiple copies (for example WPF is likely to have issues if two copies are running in the same process). There's also the problem of sharing. Reasonable expectation for frameworks is that they are "shared" by all components in the app. That would require us to load the framework into the default context (even when resolving it for the component) - which comes with lot of tricky issues.